chore: use pnpm for builds

midzelis opened the pull request.

midzelis committed:
Migrate from npm to pnpm across entire project

• Update all GitHub workflow files to use pnpm instead of npm
• Replace npm commands with pnpm equivalents in devcontainer scripts
• Remove package-lock.json files and update to use pnpm-lock.yaml
• Consolidate node version references to use server/.nvmrc

mertalev commented on server/Dockerfile:
```diff
@@ -69,31 +74,38 @@ RUN sudo apt-get update \
+WORKDIR /usr/src/app
+COPY ./package* ./pnpm* .pnpmfile.cjs ./
+COPY ./server ./server/
+RUN SHARP_IGNORE_GLOBAL_LIBVIPS=true pnpm --filter immich --frozen-lockfile build && \
```
Why is that env set?

midzelis committed:
Refine pnpm migration based on review feedback

• Replace SKIP_SHARP_FILTERING with SHARP_IGNORE_GLOBAL_LIBVIPS environment variable
• Improve Sharp package filtering to include specific Linux architectures for Docker builds
• Optimize Dockerfile dependency caching with improved layer structure
• Clean up workspace configuration and remove redundant settings

midzelis commented on open-api/bin/generate-open-api.sh:
```diff
@@ -28,11 +28,17 @@ function dart {
+  pnpm --filter immich install
+  pnpm --filter immich build
+  pnpm --filter immich sync:open-api
+)
```
Yes - the subsequent commands in the bash script don't expect the directory to be changed. By running it within a `( )` subshell, the `CWD` is restored after the subshell ends `)`

mertalev reviewed (approved):
LGTM, but I think it should get more eyes before merge

jrasm91 commented on server/src/constants.ts:
```diff
@@ -41,7 +42,10 @@ export const SALT_ROUNDS = 10;
 export const IWorker = 'IWorker';
 
-const { version } = JSON.parse(readFileSync('./package.json', 'utf8'));
+// eslint-disable-next-line unicorn/prefer-module
+const basePath = dirname(__filename);
+const packageFile = join(basePath, '..', 'package.json');
+const { version } = JSON.parse(readFileSync(packageFile, 'utf8'));
```
I think because we used to get the version from the package.json file and that doesn't exist in the server directory anymore.

jrasm91 replied:
Oh nevermind, I think it originally was removed, but it is still there in this version of the PR.

danieldietzler replied:
What is the CWD now?

danieldietzler replied:
I see. Fair enough, thanks for explaining

midzelis committed:
Address additional review feedback for pnpm migration

• Fix node-version-file paths in GitHub workflow configurations
• Refactor .pnpmfile.cjs to use switch statement for better code organization
• Correct cache type typo in fix-format workflow
• Simplify Vite configuration by merging configs inline
• Update package description for consistency

midzelis commented on docker/docker-compose.dev.yml:
```diff
@@ -21,18 +21,28 @@ services:
       - ${UPLOAD_LOCATION}/photos/upload:/data/upload
-      - /usr/src/app/server/node_modules
       - /etc/localtime:/etc/localtime:ro
+      - pnpm-store:/usr/src/app/.pnpm-store
+      - server-node-modules:/usr/src/app/server/node_modules
+      - github-node_modules:/usr/src/app/.github/node_modules
+      - cli-node_modules:/usr/src/app/cli/node_modules
+      - docs-node_modules:/usr/src/app/docs/node_modules
+      - e2e-node_modules:/usr/src/app/e2e/node_modules
+      - sdk-node_modules:/usr/src/app/open-api/typescript-sdk/node_modules
+      - app-node_modules:/usr/src/app/node_modules
+      - web-node_modules:/usr/src/app/web/node_modules
+      - sveltekit:/usr/src/app/web/.svelte-kit
+      - coverage:/usr/src/app/web/coverage
```
A few reasons: 
1) This is to avoid permissions issues between the view of the files inside the container vs the ones on the bind-mount on the host. If the container were to be run by a user other than the user that owns the root of the git-clone on the host, then the files would be created with the owner of the user running the container, and if that happened to be 'root' then the files can't be cleaned up with a simple `make clean` command - it would require `sudo make clean` which is a little more dangerous/annoying. 
2) Isolation: All of dependency/package install commands are performed from within a docker container, which may be a different architecture (MacOS on arm) or libc (the web container uses alpine with *musl* instead of *glibc*) - and some of the dependent packages use native binary code, which only make sense in the architecture/libc version that is within the container it was executed inside. By using volumes, these dependencies are not created on the hosts filesystem. 
2) pnpm performs performs a clone, hard-link, or copy of files (in that order) from the content-addressable-store (`/buildcache/pnpm-store`) to the virtual store (`.pnpm-store`). And then creates the dependency node structures in `node_modules` using symlinks. Using volume mounts makes it more likely that hardlinks will be successful, increasing speed to install dependencies.

midzelis commented on web/vite.config.ts:
```diff
@@ -60,3 +65,5 @@ export default defineConfig({
   },
 });
+
+export default mergeConfig(viteConfig, vitestConfig);
```
I'm not sure why (maybe because pnpm is more strict?), but I was not able to specify the 'test' property within a `defineConfig` from `vite`. I was able to specify `test` when importing `defineConfig` from `vitest/config` - however, then there was an type-error on the `plugins` property in the config. The one way that worked was what you see here - import both the vitest config and the vite config, then merge them. Also, the triple slash command in the docs did not help. 

In fact, I looked at all the 'solutions' here: https://stackoverflow.com/questions/72146352/vitest-defineconfig-test-does-not-exist-in-type-userconfigexport and only the `mergeConfig` technique (one of the alternatives in the comments) was the only one that worked for me.

midzelis replied:
Update: I found another solution that just needs a new type import that is a little cleaner

danieldietzler replied:
That looks good to me, thanks!

midzelis commented on .github/workflows/test.yml:
```diff
@@ -327,41 +288,35 @@ jobs:
-          cache: 'npm'
-          cache-dependency-path: '**/package-lock.json'
-
+          node-version-file: './server/.nvmrc'
```
This was a poorly executed regexp find/replace - fixed

danieldietzler commented on .github/workflows/fix-format.yml:
```diff
@@ -31,9 +31,9 @@ jobs:
-          node-version-file: './server/.nvmrc'
-          cache: 'npm'
-          cache-dependency-path: '**/package-lock.json'
+          node-version-file: './.github/.nvmrc'
```
Isn't this still wrong?

midzelis replied:
Looking more carefully, this is actually formatting all of the projects, not just .github. You'd technically have to run a version of setup-node for each project. This all may be a moot point, since the node version is always the same for every project. 

Would it make sense to just treat the node version as a workspace level setting, instead of per-project? Then, the nvmrc file could be moved up to the root directory of the git repo.

midzelis committed:
Use 'server/.nvmrc' for fix-format.yml GHA

midzelis committed:
Delete npm locks

danieldietzler commented on docker/docker-compose.dev.yml:
```diff
@@ -157,7 +160,26 @@ services:
   model-cache:
   prometheus-data:
   grafana-data:
+  pnpm-store:
+  github-node_modules:
+  server-node-modules:
+  cli-node_modules:
+  docs-node_modules:
+  e2e-node_modules:
+  sdk-node_modules:
+  app-node_modules:
+  web-node_modules:
+  sveltekit:
+  coverage:
```
Those aren't used anymore are they?

danieldietzler reviewed (approved):
Nice! Excited for pnpm

jrasm91 commented:
I don't see any big issues but I would like to test it a bit before merging it as well

midzelis commented on .devcontainer/server/container-start-backend.sh:
```diff
@@ -3,15 +3,20 @@
 log ""
 cd "${IMMICH_WORKSPACE}/server" || (
-    log "Immich workspace not found"
+    log "Immich workspace not found"jj
```
Not sure how that got there - removed

midzelis committed:
Remove Docker volume isolation for node_modules directories

• Remove volume mounts for node_modules and related directories
• Allow shared access between host and container filesystem
• Update init container to handle file ownership with conditional existence check

midzelis committed:
Remove unused Docker volumes and volume mounts

• Remove node_modules volume mounts from devcontainer configuration
• Remove unused named volumes for pnpm-store, node_modules, and cache directories
• Clean up Docker Compose configuration after removing volume isolation

midzelis committed:
Fix typescript-sdk package issues

• Remove unknown "build" dependency that was incorrectly added to package.json
• Update pnpm-lock.yaml to reflect dependency removal

midzelis committed:
Add pnpm setup to mobile workflow for translation formatting

• Add pnpm action setup step to mobile unit tests workflow
• Required for translation file formatting and sorting operations

zackpollard reviewed (approved):
These changes look good, did a bunch of testing with the dev and production server as well as various commands via pnpm and everything seems to be working well!

jrasm91 committed:
chore: merge main
